### ISSUE TITLE
PUT Requests storing empty JSON object

### ISSUE THREAD

**Issue by boulderwebdev** [solid/node-solid-server#1165]
I am trying to update text files on my pod using a `PUT` request. If I have a text file, e.g. `/public/hello.txt` and make a `PUT` request to override its value, the file is replaced by a text file with an empty JSON object.

Here's a snipped of the code I used to perform the `PUT` request. Here we are assuming `/public/hello.txt` exists already.
```javascript
url = "https://boulderwebdev.solid.community/public/hello.txt"
solid.fetch(url, {
  method: "PUT",
  headers: {
    "Content-Type": "text/json"
  },
  body: JSON.stringify({"hello": "world"})
}).then(resp => {...})
```
If you go to `https://boulderwebdev.solid.community/public/hello.txt` now you will see a page with the text
```
{}
```

**Comment by jaxoncreed** [comment]
There was a similar issue recently at #1011. It's caused by the auth system using body-parser while the ldp expects a stream. To rectify this, we added a temporary hack that creates a new stream from the parsed body, but this hack only applies to `application/json` not `text/json` and only in a `POST` request.

https://github.com/solid/node-solid-server/blob/master/lib/ldp.js#L167

If you would like, you can open a pull request that extends this hack to apply to your use case.

**Comment by boulderwebdev** [comment]
Oh, whoops, that should be `text/plain`. Thanks for the reference, I'll check it out soon.

**Comment by Otto-AA** [comment]
I've tested it a bit and from what I've seen, v4.x (solid.community) works if slug is set to "filename.json" and the Content-Type is unset, and v5.0.0 works if slug is set to "filename" and Content-Type to "application/json".
Both don't work the other way round (v4.x doesn't work with application/json and v5.0.0 requires application/json), so I haven't figured out how I should handle it to support both versions. Is there a way to do this?

And another question (I can move it to another issue if you want): Should the server accept an undefined/unset Content-Type? As far as I've seen it works perfectly to add the extension to the slug (`slug: test.css`) and don't set the Content-Type for v4.x and (except for json files) also for v5.0.0.
Is this behavior intended, and if yes, why doesn't this work for `slug: test.json` in v5.0.0?

**Comment by jaxoncreed** [comment]
No, it's not intended. This behaviour is the result of a hack located here: https://github.com/solid/node-solid-server/blob/master/lib/ldp.js#L159. We will work on fixing this, but pull requests are appreciated if you'd like to take a crack at it 😃

**Comment by Otto-AA** [comment]
Do you mean, that using `slug: filename.ext` but leaving Content-Type unset is something, that _should_ work? I myself didn't find any documentation on this, so I would be glad if you know more about this :)
(And I'll likely don't have time to get familiar with this project in the near future, sorry)

**Comment by jaxoncreed** [comment]
Maybe @michielbdejong can correct me if I'm wrong, but I believe that the file extension supersedes the content type. So if I were to submit `file.txt` with `content-type: application/json` it should still be saved in the system as `file.txt` and not `file.json`

**Comment by kjetilk** [comment]
> Maybe @michielbdejong can correct me if I'm wrong, but I believe that the file extension supersedes the content type. So if I were to submit `file.txt` with `content-type: application/json` it should still be saved in the system as `file.txt` and not `file.json`

Hmmm, in that case, I'd have to hold my nose... But perhaps there is a history behind it that I'm not aware of?

**Comment by Otto-AA** [comment]
I've taken a look at the source code: slug and Content-Type are treated equally. The Content-Type header gets converted to a file extension and then they are concatenated to the path:
https://github.com/solid/node-solid-server/blob/19d1bf0ff5a9a59bb59300b8fce3bfcd038d0ea7/lib/ldp.js#L454-L455
So for instance POST `slug: test.txt` `Content-Type: text/css` would be stored as `test.txt.css`.

But what I initially wanted to know is, if this behavior should be the same for all solid pod implementations, or in other words, if it is defined in some spec from solid or ldp. Because I would like to stick to the specification as much as possible to make sure it works with other server implementations too.

**Comment by kjetilk** [comment]
Well, it goes back to [Web Architecture](https://www.w3.org/TR/webarch/Overview.html#metadata-inconsistencies)

**Comment by timbl** [comment]
> Maybe @michielbdejong can correct me if I'm wrong, but I believe that the file extension supersedes the content type. So if I were to submit `file.txt` with `content-type: application/json` it should still be saved in the system as `file.txt` and not `file.json`

No absolutely not. A PUT without a content-type is an invalid request.

**Comment by kjetilk** [comment]
Right, so it indeed it seems like it is bypassing the `ResourceMapper`, which should follow [this design issue](https://www.w3.org/DesignIssues/HTTPFilenameMapping.html)

**Comment by jaxoncreed** [comment]
Okay thanks for the clarification!
